File: compat-android/src/com/vekom/padelprobe/MatchEngine.java
package com.vekom.padelprobe;

import java.util.ArrayList;
import java.util.List;

import static com.vekom.padelprobe.MatchModel.Mode;
import static com.vekom.padelprobe.MatchModel.Settings;
import static com.vekom.padelprobe.MatchModel.State;
import static com.vekom.padelprobe.MatchModel.Team;

final class MatchEngine {
    static final long DEBOUNCE_MS = 350L;

    static final class Result {
        final boolean accepted;
        final boolean completedNow;

        Result(boolean accepted, boolean completedNow) {
            this.accepted = accepted;
            this.completedNow = completedNow;
        }
    }

    private State state;
    private final List<State> history;
    private ScoringStrategy strategy;
    private long lastAcceptedPointAt = -1L;

    MatchEngine(Mode mode, Settings settings) {
        state = State.initial(mode, settings);
        history = new ArrayList<>();
        strategy = ScoringStrategy.Factory.forMode(mode);
    }

    static MatchEngine restore(State restored, List<State> restoredHistory) {
        MatchEngine engine = new MatchEngine(restored.mode, restored.settings);
        engine.state = restored.copy();
        engine.history.clear();
        for (State previous : restoredHistory) {
            engine.history.add(previous.copy());
        }
        engine.strategy = ScoringStrategy.Factory.forMode(restored.mode);
        return engine;
    }

    State state() {
        return state.copy();
    }

    List<State> history() {
        List<State> result = new ArrayList<>();
        for (State previous : history) {
            result.add(previous.copy());
        }
        return result;
    }

    boolean canUndo() {
        return !history.isEmpty();
    }

    Result point(Team team, long occurredAt) {
        if (state.completed) {
            return new Result(false, false);
        }
        if (lastAcceptedPointAt >= 0
                && occurredAt >= lastAcceptedPointAt
                && occurredAt - lastAcceptedPointAt < DEBOUNCE_MS) {
            return new Result(false, false);
        }
        history.add(state.copy());
        strategy.addPoint(state, team);
        state.revision++;
        lastAcceptedPointAt = occurredAt;
        return new Result(true, state.completed);
    }

    Result point(Team team) {
        return point(team, System.currentTimeMillis());
    }

    boolean undo() {
        if (history.isEmpty()) {
            return false;
        }
        state = history.remove(history.size() - 1);
        strategy = ScoringStrategy.Factory.forMode(state.mode);
        lastAcceptedPointAt = -1L;
        return true;
    }

    void changeServer() {
        history.add(state.copy());
        state.currentServer = state.currentServer.other();
        state.pointsSinceServerChange = 0;
        state.revision++;
    }

    void resetCurrent() {
        State previous = state;
        state = State.initial(previous.mode, previous.settings);
        if (previous.mode == Mode.AMERICANO) {
            state.roundNumber = previous.roundNumber;
            state.sessionPointsA = previous.sessionPointsA;
            state.sessionPointsB = previous.sessionPointsB;
            for (MatchModel.RoundResult result : previous.roundHistory) {
                state.roundHistory.add(result.copy());
            }
        }
        history.clear();
        lastAcceptedPointAt = -1L;
    }

    void clearSession() {
        state = State.initial(state.mode, state.settings);
        history.clear();
        lastAcceptedPointAt = -1L;
    }

    boolean startNextRound() {
        if (state.mode != Mode.AMERICANO || !state.completed) {
            return false;
        }
        State previous = state;
        State next = State.initial(Mode.AMERICANO, previous.settings);
        next.roundNumber = previous.roundNumber + 1;
        next.sessionPointsA = previous.sessionPointsA;
        next.sessionPointsB = previous.sessionPointsB;
        next.currentServer = previous.currentServer;
        next.revision = previous.revision + 1;
        for (MatchModel.RoundResult result : previous.roundHistory) {
            next.roundHistory.add(result.copy());
        }
        state = next;
        history.clear();
        lastAcceptedPointAt = -1L;
        return true;
    }

    String score(Team team) {
        return strategy.scoreLabel(state, team);
    }

    String modeLabel() {
        return strategy.modeLabel(state);
    }

    String statusLabel() {
        if (!state.completed) {
            return state.inTieBreak ? "TIE-BREAK" : "IN PLAY";
        }
        return "DRAW".equals(state.winner) ? "ROUND DRAW" : "TEAM " + state.winner + " WINS";
    }

    String detailLabel() {
        if (state.mode == Mode.CLASSIC || state.mode == Mode.SINGLE_SET) {
            return "SETS " + state.setsA + ":" + state.setsB
                    + "   GAMES " + state.gamesA + ":" + state.gamesB;
        }
        if (state.mode == Mode.AMERICANO) {
            int played = state.pointsA + state.pointsB;
            int remaining = Math.max(0, state.settings.target - played);
            return "ROUND " + state.roundNumber + "   LEFT " + remaining
                    + (state.settings.trackServe ? "   SERVE " + state.currentServer : "");
        }
        int remaining = Math.max(0, state.settings.target - Math.max(state.pointsA, state.pointsB));
        return "LEFT " + remaining;
    }
}

File: entry/src/main/js/MainAbility/engine/MatchEngine.js
import { createInitialState } from './Defaults';
import { AmericanoScoringStrategy, ClassicScoringStrategy, RaceToNScoringStrategy, TieBreakScoringStrategy } from './Strategies';
import { DEFAULT_DEBOUNCE_MS, SAVE_FORMAT_VERSION, cloneState, otherTeam } from './Types';
function strategyFor(mode) {
    if (mode === 'classic' || mode === 'single_set') {
        return new ClassicScoringStrategy();
    }
    if (mode === 'tie_break' || mode === 'super_tie_break') {
        return new TieBreakScoringStrategy();
    }
    if (mode === 'race_to_n') {
        return new RaceToNScoringStrategy();
    }
    return new AmericanoScoringStrategy();
}
function isValidState(value) {
    return (value !== null &&
        typeof value === 'object' &&
        value.formatVersion === SAVE_FORMAT_VERSION &&
        typeof value.mode === 'string' &&
        typeof value.pointsA === 'number' &&
        typeof value.pointsB === 'number' &&
        value.pointsA >= 0 &&
        value.pointsB >= 0 &&
        value.settings !== null &&
        typeof value.settings === 'object');
}
export class MatchEngine {
    constructor(mode = 'americano', settings, debounceMs = DEFAULT_DEBOUNCE_MS) {
        this.stateValue = createInitialState(mode, settings);
        this.historyValue = [];
        this.strategy = strategyFor(mode);
        this.lastAcceptedPointAt = -1;
        this.debounceMs = debounceMs;
    }
    get state() {
        return cloneState(this.stateValue);
    }
    get historyDepth() {
        return this.historyValue.length;
    }
    get presentation() {
        return this.strategy.presentation(this.stateValue, this.historyValue.length > 0);
    }
    dispatch(event) {
        if (event.type === 'PointWon') {
            return this.point(event.team, event.occurredAt);
        }
        if (event.type === 'Undo') {
            return this.undo();
        }
        if (event.type === 'Reset') {
            this.reset();
            return { accepted: true, reason: 'reset', completedNow: false };
        }
        if (event.type === 'ClearSession') {
            this.clearSession();
            return { accepted: true, reason: 'session cleared', completedNow: false };
        }
        if (event.type === 'ChangeServer') {
            this.historyValue.push(cloneState(this.stateValue));
            this.stateValue.currentServer = otherTeam(this.stateValue.currentServer);
            this.stateValue.pointsSinceServerChange = 0;
            this.stateValue.revision += 1;
            return { accepted: true, reason: 'server changed', completedNow: false };
        }
        return this.startNextRound();
    }
    point(team, occurredAt = Date.now()) {
        if (this.stateValue.completed) {
            return { accepted: false, reason: 'match is complete', completedNow: false };
        }
        if (this.lastAcceptedPointAt >= 0 &&
            occurredAt - this.lastAcceptedPointAt >= 0 &&
            occurredAt - this.lastAcceptedPointAt < this.debounceMs) {
            return { accepted: false, reason: 'debounced duplicate tap', completedNow: false };
        }
        this.historyValue.push(cloneState(this.stateValue));
        this.strategy.addPoint(this.stateValue, team);
        this.stateValue.revision += 1;
        this.lastAcceptedPointAt = occurredAt;
        return {
            accepted: true,
            reason: 'point added',
            completedNow: this.stateValue.completed
        };
    }
    undo() {
        const previous = this.historyValue.pop();
        if (previous === undefined) {
            return { accepted: false, reason: 'nothing to undo', completedNow: false };
        }
        this.stateValue = previous;
        this.strategy = strategyFor(this.stateValue.mode);
        this.lastAcceptedPointAt = -1;
        return { accepted: true, reason: 'undone', completedNow: false };
    }
    reset() {
        const mode = this.stateValue.mode;
        const settings = this.stateValue.settings;
        const roundHistory = cloneState(this.stateValue).roundHistory;
        const sessionA = this.stateValue.sessionPointsA;
        const sessionB = this.stateValue.sessionPointsB;
        const roundNumber = this.stateValue.roundNumber;
        this.stateValue = createInitialState(mode, settings);
        if (mode === 'americano') {
            this.stateValue.roundHistory = roundHistory;
            this.stateValue.sessionPointsA = sessionA;
            this.stateValue.sessionPointsB = sessionB;
            this.stateValue.roundNumber = roundNumber;
        }
        this.historyValue = [];
        this.lastAcceptedPointAt = -1;
    }
    clearSession() {
        const mode = this.stateValue.mode;
        const settings = this.stateValue.settings;
        this.stateValue = createInitialState(mode, settings);
        this.historyValue = [];
        this.lastAcceptedPointAt = -1;
    }
    startNextRound() {
        if (this.stateValue.mode !== 'americano' || !this.stateValue.completed) {
            return { accepted: false, reason: 'completed Americano round required', completedNow: false };
        }
        const previous = this.stateValue;
        const next = createInitialState('americano', previous.settings);
        next.roundNumber = previous.roundNumber + 1;
        next.roundHistory = cloneState(previous).roundHistory;
        next.sessionPointsA = previous.sessionPointsA;
        next.sessionPointsB = previous.sessionPointsB;
        next.currentServer = previous.currentServer;
        next.revision = previous.revision + 1;
        this.stateValue = next;
        this.historyValue = [];
        this.lastAcceptedPointAt = -1;
        return { accepted: true, reason: 'next round started', completedNow: false };
    }
    serialize() {
        const data = {
            formatVersion: SAVE_FORMAT_VERSION,
            state: cloneState(this.stateValue),
            history: this.historyValue.map((item) => cloneState(item)),
            lastAcceptedPointAt: this.lastAcceptedPointAt
        };
        return JSON.stringify(data);
    }
    static restore(raw, fallbackMode = 'americano') {
        try {
            const parsed = JSON.parse(raw);
            if (parsed.formatVersion !== SAVE_FORMAT_VERSION ||
                !isValidState(parsed.state) ||
                !Array.isArray(parsed.history)) {
                return {
                    engine: new MatchEngine(fallbackMode),
                    restored: false,
                    reason: 'unsupported or invalid save format'
                };
            }
            const engine = new MatchEngine(parsed.state.mode, parsed.state.settings);
            engine.stateValue = cloneState(parsed.state);
            engine.historyValue = parsed.history
                .filter((state) => isValidState(state))
                .map((state) => cloneState(state));
            engine.lastAcceptedPointAt =
                typeof parsed.lastAcceptedPointAt === 'number' ? parsed.lastAcceptedPointAt : -1;
            engine.strategy = strategyFor(parsed.state.mode);
            return { engine, restored: true, reason: 'restored' };
        }
        catch (_error) {
            return {
                engine: new MatchEngine(fallbackMode),
                restored: false,
                reason: 'corrupt save data'
            };
        }
    }
}
